File: sdk/go/matcher.go
package tuicast

// Matcher is a screen condition evaluated by the driver while waiting.
type Matcher interface {
	driverMatcher() matcherSpec
}

type matcher struct{ specification matcherSpec }

func (value matcher) driverMatcher() matcherSpec { return value.specification }

type matcherSpec struct {
	Contains *string        `json:"contains,omitempty"`
	Line     *lineMatcher   `json:"line,omitempty"`
	Cursor   *cursorMatcher `json:"cursor,omitempty"`
	All      []matcherSpec  `json:"all,omitempty"`
	Any      []matcherSpec  `json:"any,omitempty"`
	Not      *matcherSpec   `json:"not,omitempty"`
}

type lineMatcher struct {
	Row  int    `json:"row"`
	Text string `json:"text"`
}

type cursorMatcher struct {
	Column int `json:"column"`
	Row    int `json:"row"`
}

// Contains matches exact text anywhere in the complete screen text.
func Contains(text string) Matcher {
	return matcher{specification: matcherSpec{Contains: &text}}
}

// LineEquals matches one complete zero-based screen row, including trailing
// spaces.
func LineEquals(row int, text string) Matcher {
	return matcher{specification: matcherSpec{Line: &lineMatcher{Row: row, Text: text}}}
}

// CursorAt matches a zero-based cursor position.
func CursorAt(column, row int) Matcher {
	return matcher{specification: matcherSpec{Cursor: &cursorMatcher{Column: column, Row: row}}}
}

// All matches when every child matcher succeeds.
func All(matchers ...Matcher) Matcher {
	children := make([]matcherSpec, len(matchers))
	for index, child := range matchers {
		if child != nil {
			children[index] = child.driverMatcher()
		}
	}
	return matcher{specification: matcherSpec{All: children}}
}

// Any matches when at least one child matcher succeeds.
func Any(matchers ...Matcher) Matcher {
	children := make([]matcherSpec, len(matchers))
	for index, child := range matchers {
		if child != nil {
			children[index] = child.driverMatcher()
		}
	}
	return matcher{specification: matcherSpec{Any: children}}
}

// Not matches when its child matcher does not succeed.
func Not(child Matcher) Matcher {
	if child == nil {
		return matcher{}
	}
	specification := child.driverMatcher()
	return matcher{specification: matcherSpec{Not: &specification}}
}

File: sdk/java/src/main/java/dev/castingcode/tuicast/Matcher.java
package dev.castingcode.tuicast;

import java.util.*;

public sealed interface Matcher permits Matcher.Spec {
  Object value();

  record Spec(Object value) implements Matcher {}

  static Matcher contains(String text) {
    return new Spec(Map.of("contains", text));
  }

  static Matcher lineEquals(int row, String text) {
    return new Spec(Map.of("line", Map.of("row", row, "text", text)));
  }

  static Matcher cursorAt(int column, int row) {
    return new Spec(Map.of("cursor", Map.of("column", column, "row", row)));
  }

  static Matcher all(Matcher... values) {
    return composite("all", values);
  }

  static Matcher any(Matcher... values) {
    return composite("any", values);
  }

  static Matcher not(Matcher value) {
    return new Spec(Map.of("not", Objects.requireNonNull(value).value()));
  }

  private static Matcher composite(String name, Matcher[] values) {
    if (values.length == 0) throw new IllegalArgumentException(name + " requires children");
    return new Spec(
        Map.of(name, Arrays.stream(values).map(x -> Objects.requireNonNull(x).value()).toList()));
  }
}
